//
// Copyright (C) 2024 EA group inc.
// Author: Jeff.li lijippy@163.com
// All rights reserved.
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

syntax="proto2";
package sirius.proto;
option cc_generic_services = true;

import "sirius/proto/error_code.proto";

enum RaftControlOp {
  SetPeer = 1;
  TransLeader = 2;
  SnapShot = 3;
  GetLeader = 4;
  ShutDown = 5;
  Vote = 6;
  ResetVoteTime = 7;
  ListPeer = 8;
};

// operation request/response

message RaftControlRequest {
  required RaftControlOp op_type = 1;
  optional int64 region_id = 2;
  optional string new_leader = 3;
  repeated string old_peers = 4;
  repeated string new_peers = 5;
  optional int64 election_time = 6;
  optional bool force = 7;
};

message RaftControlResponse {
  required int64 region_id = 1;
  required ErrCode errcode = 2;
  optional string leader = 3;
  optional string errmsg = 4;
  repeated string peers  = 5;
};

enum RecoverOpt {
  DO_NONE                = 0;
  DO_REMOVE_ILLEGAL_PEER = 1;
  DO_REMOVE_PEER         = 2;
  DO_SET_PEER            = 3;
  DO_INIT_REGION         = 4;
};
